sqlmesh/core/engine_adapter/mssql.py

```diff
@@ -191,14 +193,23 @@ def query_factory() -> Query:
                     elif getattr(df.dtypes[column], "tz", None) is not None:  # type: ignore
                         # MSSQL requires a colon in the offset (+00:00) so we use isoformat() instead of strftime()
                         df[column] = pd.to_datetime(df[column]).map(lambda x: x.isoformat(" "))  # type: ignore
+
+                        # bulk_copy() doesn't work with TZ timestamp, so load into string column and cast to
+                        # timestamp in SELECT statement
+                        columns_to_types_tztext = columns_to_types.copy()
```

Comment:
let's just make the copy upfront on line 189, that way we aren't copying n times for every tz

Reply:
so just modify this in place?

Reply:
I see the problem, but I'm not following what you mean by modify in place. 

What if we move the dt64 detection outside the loop so we know up front whether there's any need to copy?

``` python
dt64_cols = {k:v for (k,v) in columns_to_types.items() if is_datetime64_any_dtype(df.dtypes[k])}
columns_to_types_tztext = columns_to_types.copy() if dt64_cols else None
for column, kind in dt64_cols.items():
    if kind.is_type("date"):  # type: ignore
    ...
```

Reply:
what i meant was

```python

for column, kind in (columns_to_types or {}).copy().items():

     columns_to_types[column] = Text